src/instructlab/training/main_ds.py

```diff
@@ -382,11 +454,36 @@ def train(
             if not args.use_dolomite:
                 for k in batch:
                     batch[k] = batch[k].to(local_rank)
-            output = model(
+
+            # get the training loss from running on data
+            output: CausalLMOutput = model(
                 **batch,
                 use_cache=False,
             )
-            loss = output.loss
+
+            loss = None
+            if args.distill:
+                # teacher_model should always be provided when `args.distill` is enabled
+                if TYPE_CHECKING:
+                    assert (
+                        teacher_model is not None
+                    ), "teacher model cannot be None when `distill` is enabled"
+
+                with torch.no_grad():
+                    teacher_output: CausalLMOutput = teacher_model(
+                        **batch, use_cache=False
+                    )
+                loss = distillation_loss(
+                    student_output=output,
+                    teacher_output=teacher_output,
+                    alpha=args.distill_alpha,
+                    temp=args.distill_temp,
+                )
+
+            else:
+                loss = output.loss
+
+            assert loss is not None, "loss cannot be equal to None!"
```

Comment:
`assert`s are typically not preferred in comparison to runtime exceptions.
```suggestion
            if loss is None:
                raise ValueError("loss was None during distillation training. Something unrecoverable went wrong.")
```

Reply:
mostly because they can be removed with `-0` when the interpreter is invoked. But we want to check non-null all the time.

Reply:
Sure I can change this. I was using them as scaffolding when writing this.

Reply:
I'm gonna make it not be specific to distillation training though since it's more about how we branch out. I suspect as we add other loss calculations (contrastive loss, preference tuning loss, etc.), we will start out by setting it to `None` and having this final check to ensure it was set to something.